### PR TITLE
feat(image_projection_based_fusion, roi_cluster_fusion): roi and obstacle fusion method

### PR DESCRIPTION
**Pull request by yukke42**
## Description

I plan to add a feature `roi_detected_object_fusion` to fuse ROIs from images and Detected Objects,  but the most of the implementation is the same as `roi_cluster_fusion`. Therefore, I created a new package `image_projection_based_fusion`  and it has these two packages.

`roi_cluster_fusion`: 

- The latest tf is used to transform from the frame_id of a cluster to that of an ROI. I changes to use tf at the stamp of a ROI.

`roi_detected_object_fusion`:

- The fusion algorithm is the same as `roi_cluster_fusion`, except that a classification label of a detected object doesn't have `UNKNOWN`.

## Related links

https://github.com/tier4/autoware_launch/pull/274

## Pre-review checklist for the PR author

PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

Reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has the write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by yukkysaito** [comment]
I changed to draft

File: perception/image_projection_based_fusion/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+The `image_projection_based_fusion` is a package to fuse detected obstacles (bounding box or segmentation) from image and 3d pointcloud or obstacles (bounding box, cluster or segmentation).
+
+## Inner-workings / Algorithms
+
+Detail description of each fusion's algorithm is in the following links.
+
+| Fusion Name                | Description                                                                                    | Detail                                       |
+| -------------------------- | ---------------------------------------------------------------------------------------------- | -------------------------------------------- |
+| roi_cluster_fusion         | Overwrite a classfication label of clusters by that of ROIs from a 2D object detector.         | [link](./docs/roi-cluster-fusion.md)         |
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/README.md:13]:
```suggestion
| roi_cluster_fusion         | Overwrite a classification label of clusters by that of ROIs from a 2D object detector.         | [link](./docs/roi-cluster-fusion.md)         |
```

File: perception/image_projection_based_fusion/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+| roi_detected_object_fusion | Overwrite a classfication label of detected objects by that of ROIs from a 2D object detector. | [link](./docs/roi-detected-object-fusion.md) |
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/README.md:14]:
```suggestion
| roi_detected_object_fusion | Overwrite a classification label of detected objects by that of ROIs from a 2D object detector. | [link](./docs/roi-detected-object-fusion.md) |
```

**Comment by yukkysaito** [comment]
Might take some time to review.

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi2_msg,
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi3_msg,
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi4_msg,
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi5_msg,
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi6_msg,
+    DetectedObjectsWithFeature::ConstSharedPtr input_roi7_msg);
+
+  virtual void preprocess();
+
+  virtual void fusionOnSingleImage(
```
Comment (yukkysaito) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp:67]: I want to start with a verb.

File: perception/image_projection_based_fusion/README.md (hunk 1 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/README.md]: fixed at [a074fbf](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/a074fbfe6bd1f9f10d1be2b0f792784b2a05d0aa)

File: perception/image_projection_based_fusion/README.md (hunk 2 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/README.md]: fixed at [a074fbf](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/a074fbfe6bd1f9f10d1be2b0f792784b2a05d0aa)

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp (hunk 3 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp]: changed at [3e58d25](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/3e58d25ec2c03790d6a9938bc43741128f695a6a)

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    DetectedObjectsWithFeature, DetectedObjectsWithFeature, DetectedObjectsWithFeature,
+    DetectedObjectsWithFeature, DetectedObjectsWithFeature>;
+  using Sync = message_filters::Synchronizer<SyncPolicy>;
+  typename std::shared_ptr<Sync> sync_ptr_;
+
+  // output
+  // if the type of output message isn't Msg,
+  // overwrite output_msg_ and pub_ptr_ in derived class.
+  Msg input_msg_;
+  Msg output_msg_;
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp]: changed at [585356f](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/585356f80e368476bfa09cbda3cc7e6dfb001bc8)

**Comment by yukke42** [comment]
I found that 2d-3d projeciton fails on some images, so I'll fix it.

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yukkysaito) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp]: @yukke42 it is dead link :pray:

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp (hunk 3 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp]: @yukkysaito 
sorry, changed to [ad2e762](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/ad2e762b9d0c5d636e304035526d64800d61febf)

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/fusion_node.hpp]: changed to [2b0e0d2](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/2b0e0d2d6e28c51450355d22ccc49647b7d080ff)

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  }
+}
+
+void RoiDetectedObjectFusionNode::generateDetectedObjectRois(
+  const std::vector<DetectedObject> & input_objects, const double image_width,
+  const double image_height, const Eigen::Affine3d & object2camera_affine,
+  const Eigen::Matrix4d & camera_projection,
+  std::map<std::size_t, sensor_msgs::msg::RegionOfInterest> & object_roi_map)
+{
+  for (std::size_t obj_i = 0; obj_i < input_objects.size(); obj_i++) {
```
Comment (miursh) [inline perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp:77]:
```suggestion
  for (std::size_t obj_i = 0; obj_i < input_objects.size(); ++obj_i) {
```

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      }
+      if (use_iou_x_) {
+        iou_x = calcIoUX(object_itr->second, roi);
+      }
+      if (use_iou_y_) {
+        iou_y = calcIoUY(object_itr->second, roi);
+      }
+
+      if (iou + iou_x + iou_y > max_iou) {
+        object_i = object_itr->first;
```
Comment (miursh) [inline perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp:163]: [IMO] Using range-based for statement w/ const reference.
```suggestion
    for (const auto & obj_map : object_roi_map) {
      double iou(0.0), iou_x(0.0), iou_y(0.0);
      if (use_iou_) {
        iou = calcIoU(obj_map.second, roi);
      }
      if (use_iou_x_) {
        iou_x = calcIoUX(obj_map.second, roi);
      }
      if (use_iou_y_) {
        iou_y = calcIoUY(obj_map.second, roi);
      }

      if (iou + iou_x + iou_y > max_iou) {
        object_i = obj_map.first;

```

File: perception/image_projection_based_fusion/src/fusion_node.cpp [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    RCLCPP_WARN(
+      this->get_logger(), "maximum rois_number is 8. current rois_number is %d", rois_number_);
+    rois_number_ = 8;
+  }
+
+  // subscribers
+  sub_.subscribe(this, "input", rclcpp::QoS(1).get_rmw_qos_profile());
+
+  camera_info_subs_.resize(rois_number_);
+  for (int roi_i = 0; roi_i < rois_number_; roi_i++) {
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/fusion_node.cpp]: fixed at [f53a818](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/f53a8184d543cc2a2b1086f90b94a99b6f73590b).

File: perception/image_projection_based_fusion/src/fusion_node.cpp [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    std::function<void(const sensor_msgs::msg::CameraInfo::ConstSharedPtr msg)> fnc =
+      std::bind(&FusionNode::cameraInfoCallback, this, std::placeholders::_1, roi_i);
+    camera_info_subs_.at(roi_i) = this->create_subscription<sensor_msgs::msg::CameraInfo>(
+      "input/camera_info" + std::to_string(roi_i), rclcpp::QoS{1}.best_effort(), fnc);
+  }
+
+  rois_subs_.resize(rois_number_);
+  for (int roi_i = 0; roi_i < rois_number_; roi_i++) {
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/fusion_node.cpp]: fixed at [f53a818](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/f53a8184d543cc2a2b1086f90b94a99b6f73590b).

File: perception/image_projection_based_fusion/src/fusion_node.cpp [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+{
+  if (pub_ptr_->get_subscription_count() < 1) {
+    return;
+  }
+
+  Msg output_msg = *input_msg;
+
+  preprocess(output_msg);
+
+  for (std::size_t image_id = 0; image_id < rois_subs_.size(); image_id++) {
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/fusion_node.cpp]: fixed at [f53a818](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/f53a8184d543cc2a2b1086f90b94a99b6f73590b).

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  generateDetectedObjectRois(
+    input_object_msg.objects, static_cast<double>(camera_info.width),
+    static_cast<double>(camera_info.height), object2camera_affine, camera_projection,
+    object_roi_map);
+  updateDetectedObjectClassification(
+    input_roi_msg.feature_objects, object_roi_map, output_object_msg.objects);
+
+  if (debugger_) {
+    debugger_->image_rois_.reserve(input_roi_msg.feature_objects.size());
+    for (std::size_t roi_i = 0; roi_i < input_roi_msg.feature_objects.size(); roi_i++) {
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp]: fixed at [f53a818](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/f53a8184d543cc2a2b1086f90b94a99b6f73590b).

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp (hunk 5 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp]: fixed at [f53a818](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/f53a8184d543cc2a2b1086f90b94a99b6f73590b).

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp (hunk 6 above)
Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp]: changed at [a33f315](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/a33f315f5c05d6d651483cb2c376d8eaa865897a).

File: perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    if (
+      iou_threshold_ < max_iou &&
+      output_cluster_msg.feature_objects.at(index).object.existence_probability <=
+        input_roi_msg.feature_objects.at(i).object.existence_probability &&
+      input_roi_msg.feature_objects.at(i).object.classification.front().label !=
+        autoware_auto_perception_msgs::msg::ObjectClassification::UNKNOWN) {
+      output_cluster_msg.feature_objects.at(index).object.classification =
+        input_roi_msg.feature_objects.at(i).object.classification;
+    }
+    debug_image_rois.push_back(input_roi_msg.feature_objects.at(i).feature.roi);
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_cluster_fusion/node.cpp]: changed at [a33f315](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/a33f315f5c05d6d651483cb2c376d8eaa865897a).

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+  }
+}
+
+void RoiDetectedObjectFusionNode::updateDetectedObjectClassification(
+  const std::vector<DetectedObjectWithFeature> & image_rois,
+  const std::map<std::size_t, sensor_msgs::msg::RegionOfInterest> & object_roi_map,
+  std::vector<DetectedObject> & output_objects)
+{
+  for (std::size_t roi_i = 0; roi_i < image_rois.size(); roi_i++) {
+    const auto & roi = image_rois.at(roi_i).feature.roi;
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp]: changed at [a33f315](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/a33f315f5c05d6d651483cb2c376d8eaa865897a).

File: perception/image_projection_based_fusion/package.xml [hunk 13]
```diff
@@ -1,13 +1,15 @@
...
-<package format="2">
-  <name>roi_cluster_fusion</name>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>image_projection_based_fusion</name>
   <version>0.1.0</version>
-  <description>The roi_cluster_fusion package</description>
+  <description>The image_projection_based_fusion package</description>
 
   <maintainer email="yukihiro.saito@tier4.jp">Yukihiro Saito</maintainer>
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/package.xml:8]: @yukkysaito Should I be a maintainer?

**Comment by yukke42** [comment]
The PR to launcher is created at https://github.com/tier4/autoware_launch/pull/274

File: perception/image_projection_based_fusion/package.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (yukkysaito) [reply perception/image_projection_based_fusion/package.xml:8]:
```suggestion
  <maintainer email="yukihiro.saito@tier4.jp">Yukihiro Saito</maintainer>
  <maintainer email="muramatsu.yusuke@tier4.jp">Yusuke Muramatsu</maintainer>
```

[the remark this answers is not in the document]
Comment (kenji-miyake) [reply perception/image_projection_based_fusion/package.xml:8]: muramatsu.yusuke@tier4.jp -> yusuke.muramatsu@tier4.jp?

[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/package.xml:8]: changed at [d18f43b](https://github.com/autowarefoundation/autoware.universe/pull/548/commits/d18f43b4f8785fd4642ced2bbfc2ece666462d87)

**Review by miursh — APPROVED** [review]
LGTM

**Pull request merged**